### ISSUE TITLE
Incremental parsing falls out of sync

### ISSUE THREAD

**Issue by Jollywatt** [typst/webapp-issues#177]
### Description

I think this is an issue that keeps popping up, but is hard to reproduce and pin down: sometimes, what is in the editor falls out of step with what the compiler and syntax highlighter seem to “see”. See also:
- https://github.com/typst/webapp-issues/issues/149
- https://github.com/typst/webapp-issues/issues/112
- https://github.com/typst/webapp-issues/issues/43
- https://discord.com/channels/1054443721975922748/1093454472363130961 “Button to trigger re-compilation”

Here are steps to reproduce an example of this kind. (Web app version 2023-10-19T10:29:31.555Z, compiler version e546dea2379999a934a6d402a7c56ae4478c3114)

1. Start with this file in the web app:

```typ
#let i = 1
#while i <= 10000 {
  box([#i], stroke: 1pt, inset: 2pt)
  i += 1
}
```

2. Quickly delete the `[#i], ` one character at a time.

<img width="308" alt="Screen Shot 2023-10-22 at 9 51 58 AM" src="https://github.com/typst/webapp-issues/assets/24970860/2c7d8de0-3acc-4acd-903f-aa12323a7e09">

If you don’t see the syntax highlighting fall out of sync like above, undo and try again, deleting one-by-one faster or with a larger upper limit on the while loop. It seems slip out of sync when many edits occur before the incremental compiler can catch up.

Once out of sync, always out of sync: you can't seem to get rid of the corrupt state by commenting or deleting lines:

<img width="381" alt="Screen Shot 2023-10-22 at 10 00 23 AM" src="https://github.com/typst/webapp-issues/assets/24970860/2352406b-ef32-4021-a992-dad64a4953d2">

I was also able to observe this by quickly deleting the `lorem(50),` argument character-by-character in:
```typ
#circle(
  inset: 0pt,
  rect(
    fill: red,
    inset: 0pt,
    circle(
      lorem(50),
      fill: white,
      inset: 0pt,
      radius: 4cm,
    )
  )
)
```

I would guess that the issues I linked are also examples of this sort of error. Hopefully the devs can study this example under a microscope and finally pin it!

### Browsers

Chrome, Safari

### OS

macOS

**Comment by laurmaedje** [comment]
Wow, I could reproduce it first try. Thank you for this! Having a reproduction is extremely helpful. Over the past few weeks, I have largely rewritten the web app's core and refactored large parts of the web app in general. I will study the example to ensure that the rewritten version doesn't suffer from the same issue.

**Comment by laurmaedje** [comment]
Good news: I just tested with my dev version and the problem does not occur there. I'll still try to debug the old version to find out what the cause was.

**Comment by laurmaedje** [comment]
Should be fixed since 0.9.
